x-pack/apm-server/aggregation/servicemetrics/aggregator_test.go

```diff
@@ -0,0 +1,344 @@
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zaptest/observer"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/apm-server/internal/model"
+)
+
+func TestNewAggregatorConfigInvalid(t *testing.T) {
+	report := makeErrBatchProcessor(nil)
+
+	type test struct {
+		config AggregatorConfig
+		err    string
+	}
+
+	for _, test := range []test{{
+		config: AggregatorConfig{},
+		err:    "BatchProcessor unspecified",
+	}, {
+		config: AggregatorConfig{
+			BatchProcessor: report,
+		},
+		err: "MaxGroups unspecified or negative",
+	}, {
+		config: AggregatorConfig{
+			BatchProcessor: report,
+			MaxGroups:      1,
+		},
+		err: "Interval unspecified or negative",
+	}} {
+		agg, err := NewAggregator(test.config)
+		require.Error(t, err)
+		require.Nil(t, agg)
+		assert.EqualError(t, err, "invalid aggregator config: "+test.err)
+	}
+}
+
+func TestAggregatorRun(t *testing.T) {
+	batches := make(chan model.Batch, 1)
+	agg, err := NewAggregator(AggregatorConfig{
+		BatchProcessor: makeChanBatchProcessor(batches),
+		Interval:       1 * time.Millisecond,
+		MaxGroups:      1000,
+	})
+	require.NoError(t, err)
+
+	type input struct {
+		serviceName        string
+		serviceEnvironment string
+		agentName          string
+		transactionType    string
+		outcome            string
+		count              float64
+	}
+
+	inputs := []input{
+		{serviceName: "ignored", agentName: "ignored", transactionType: "ignored", count: 0}, // ignored because count is zero
+
+		{serviceName: "backend", agentName: "java", transactionType: "request", outcome: "success", count: 2},
+		{serviceName: "backend", agentName: "java", transactionType: "request", outcome: "failure", count: 3},
+		{serviceName: "backend", agentName: "java", transactionType: "request", outcome: "unknown", count: 1},
+
+		{serviceName: "backend", agentName: "go", transactionType: "request", outcome: "unknown", count: 1},
+		{serviceName: "backend", agentName: "go", transactionType: "background", outcome: "unknown", count: 1},
+
+		{serviceName: "frontend", agentName: "rum-js", transactionType: "page-load", outcome: "unknown", count: 1},
+		{serviceName: "frontend", serviceEnvironment: "staging", agentName: "rum-js", transactionType: "page-load", outcome: "unknown", count: 1},
+	}
+
+	var wg sync.WaitGroup
+	for _, in := range inputs {
+		wg.Add(1)
+		go func(in input) {
+			defer wg.Done()
+			transaction := makeTransaction(
+				in.serviceName, in.agentName, in.transactionType,
+				in.outcome, time.Millisecond, in.count,
+			)
+			transaction.Service.Environment = in.serviceEnvironment
+
+			batch := model.Batch{transaction}
+			err := agg.ProcessBatch(context.Background(), &batch)
+			require.NoError(t, err)
+			assert.Equal(t, model.Batch{transaction}, batch)
+		}(in)
+	}
+	wg.Wait()
+
+	// Start the aggregator after processing to ensure metrics are aggregated deterministically.
+	go agg.Run()
+	defer agg.Stop(context.Background())
```

Comment:
```suggestion
	t.Cleanup(func() { agg.Stop(context.Background()) })
```

Reply:
What's the benefit? I tend to use Cleanup when creating resources outside of the main test function, which should be disposed of when the test ends.

Reply:
I suppose there's no huge benefit in this case, but since we're writing "new" tests for this aggregator, I thought we may as well use `t.Cleanup`, but it doesn't really matter.

Reply:
OK. I'm going to leave it as it is, as there's no immediate benefit. Maybe we can take a look later and see if there would be some benefit in having a test helper function for creating the aggregator, and stopping it on cleanup.